backends/p4tools/testgen/core/small_step/table_stepper.h

```diff
@@ -0,0 +1,207 @@
+    /// a P4C_UNIMPLEMENTED is thrown.
+    virtual const IR::Expression* computeTargetMatchType(
+        ExecutionState* nextState, const KeyProperties& keyProperties,
+        std::map<cstring, const FieldMatch>* matches, const IR::Expression* hitCondition);
+
+    /// A helper function that computes whether a control-plane/table-key hits or not. This does not
+    /// handle constant entries, it is specialized for control plane entries.
+    /// The function also tracks the list of field matches created to achieve a  hit. We later use
+    /// this to insert table entries using the STF/PTF framework.
+    const IR::Expression* computeHit(ExecutionState* nextState, const IR::P4Table* table,
+                                     std::map<cstring, const FieldMatch>* matches);
+
+    /// Collects properties that may be set per table. Target back end may have different semantics
+    /// for table execution that need to be collect before evaluation the table.
+    virtual void checkTargetProperties(
+        const std::vector<const IR::ActionListElement*>& tableActionList);
+
+    /* =========================================================================================
+     * Table Evaluation functions
+     * ========================================================================================= */
+    /// When we hit a table.apply() method call expression we transition into the table call
+    /// phase. Handling table calls is fairly involved as we need to respect the control plane
+    /// manipulating these tables. This is why we are using several helper functions.
+    /// @param table the table we invoke.
+    void evalTableCall();
+
+    /// Tables may have constant match-action entries. This usually implies that the table is
+    /// immutable. This helper function unrolls all entries and aligns the table key with the match
+    /// provided by the constant entry.
+    /// @param table the table we invoke.
+    /// @returns tableMissCondition reference to the current constraints of the miss condition.
+    /// Any constant entry we hit implies that the table is hit, and the default action is not
+    /// executed.
+    const IR::Expression* evalTableConstEntries();
+
+    /// This helper function evaluates potential insertion from the control plane. We use zombie
+    /// variables to mimic an operator inserting entries. We only cover ONE entry per table for
+    /// now.
+    /// @param table the table we invoke.
+    void evalTableControlEntries(const std::vector<const IR::ActionListElement*>& tableActionList);
+
+    /// This is a special function that handles the case where the key of a table is tainted. This
+    /// means that the entire execution of the table is tainted. We can dramatically simplify
+    /// execution here and we also do not need control plane entries.
+    /// The tricky part is that we still need execute all possible actions and taint whatever they
+    /// may write, iff the table has constant entries.
```

Comment:
why only if they are constant entries?

Reply:
If the table keys are tainted and the table does not have constant entries, we simply do not insert a table entry and execute the default action, which is a safe choice.